lib/common/client/base.rb

```diff
@@ -0,0 +1,152 @@
+    end
+
+    ##
+    # Base class for creating HTTP services. Wraps the Faraday gem and is configured via by passing in a
+    # {Common::Client::Configuration::REST} or {Common::Client::Configuration::SOAP} depending on the type
+    # of service you're connecting to. Once configured requests are made via the `perform` method.
+    #
+    # @example Create a service and make a GET request
+    #   class MyService < Common::Client::Base
+    #     configuration MyConfiguration
+    #
+    #     def get_resource
+    #       perform(:get, '/api/v1/resource')
+    #     end
+    #   end
+    #
+    #   service = MyService.new
+    #   response = service.get_resource
+    #
+    # @example a POST request with a body, headers, and Faraday options
+    #   def post_resource(json)
+    #     headers = { 'Content-Type' => 'application/json' }
+    #     options = { timeout: 60 }
+    #     response = perform(:post, '/submit', json, headers, options)
+    #   end
+    #
+    class Base
+      ##
+      # Sets the configuration singleton to use
+      #
+      def self.configuration(configuration = nil)
+        @configuration ||= configuration.instance
+      end
+
+      private
+
+      def config
+        self.class.configuration
+      end
+
+      def connection
+        @connection ||= lambda do
+          connection = config.connection
+          # handlers = connection.builder.handlers
+
+          # not using HTTPClient
+          # if handlers.include?(Faraday::Adapter::HTTPClient) &&
+          #   !handlers.include?(Common::Client::Middleware::Request::RemoveCookies)
+          #  raise SecurityError, 'http client needs cookies stripped'
+          # end
+
+          # not using breakers for Redis at this time
+          # if handlers.include?(Breakers::UptimeMiddleware)
+          #  return connection if handlers.first == Breakers::UptimeMiddleware
+          #
+          #  raise BreakersImplementationError, 'Breakers should be the first middleware implemented.'
+          # else
+          #  warn("Breakers is not implemented for service: #{config.service_name}")
+          # end
+
+          connection
+        end.call
+      end
+
+      def perform(method, path, params, headers = nil, options = nil)
+        raise NoMethodError, "#{method} not implemented" unless config.request_types.include?(method)
+
+        send(method, path, params || {}, headers || {}, options || {})
+      end
+
+      def request(method, path, params = {}, headers = {}, options = {}) # rubocop:disable Metrics/MethodLength
+        sanitize_headers!(method, path, params, headers)
+        raise_not_authenticated if headers.keys.include?('Token') && headers['Token'].blank?
+        connection.send(method.to_sym, path, params) do |request|
+          request.headers.update(headers)
+          options.each { |option, value| request.options.send("#{option}=", value) }
+        end.env
+      rescue Common::Exceptions::BackendServiceException => e
```

Comment:
Can you explain the reason you need to coerce your API's service exceptions?

Reply:
This is same as behavior at https://github.com/department-of-veterans-affairs/vets-api/blob/master/lib/common/client/base.rb#L107

This specific rescue creates or uses a ServiceException in the module of class making the request which allows for easier debugging when an issue occurs as the exception then appears to be coming from the correct module instead of a generic one